src/librustc_interface/queries.rs

```diff
@@ -283,46 +283,31 @@ impl<'tcx> Queries<'tcx> {
     }
 
     pub fn linker(&'tcx self) -> Result<Linker> {
-        let dep_graph = self.dep_graph()?;
-        let prepare_outputs = self.prepare_outputs()?;
-        let ongoing_codegen = self.ongoing_codegen()?;
+        let dep_graph = self.dep_graph()?.peek().clone();
+        let ongoing_codegen = self.ongoing_codegen()?.take();
+        let prepare_outputs = self.prepare_outputs()?.take();
 
         let sess = self.session().clone();
         let codegen_backend = self.codegen_backend().clone();
 
-        Ok(Linker {
-            sess,
-            dep_graph: dep_graph.peek().clone(),
-            prepare_outputs: prepare_outputs.take(),
-            ongoing_codegen: ongoing_codegen.take(),
-            codegen_backend,
-        })
+        let codegen_results = codegen_backend.join_codegen(ongoing_codegen, &sess, &dep_graph)?;
```

Comment:
This is introducing a stall inside what looks _kind of_ like an accessor, when there wasn't one before. I think there should be a comment on the function, at least. Also, double check that this doesn't seem like a problem at any call sites.

Reply:
Good point, and after a second thought, I think it makes more sense to keep `join_codegen` in where `join_codegen_and_link` was, so that behavior changes should be minimized.

I add a commit to the PR moving `join_codegen` into `Linker::link` to demonstrate the change, could you take a look?

Reply:
BTW, we can still drop `dep_graph` before calling backend to link, so we should reduce similar memory consumption during linking, as if we join codegen inside `Queries::linker()`.